### PR TITLE
Add did_message_spend API

### PR DESCRIPTION
**Pull request by ytx1991**
https://github.com/orgs/Chia-Network/projects/13/views/21

### CODE REVIEW

**Comment by ytx1991** [comment]
> @ytx1991 DID tests are failing

Transient issue

File: chia/rpc/wallet_rpc_api.py [hunk 1]
```diff
@@ -1685,6 +1687,74 @@ async def did_message_spend(self, request) -> EndpointResult:
...
+        # Get coin state
+        peer: Optional[WSChiaConnection] = self.service.get_full_node_peer()
+        if peer is None:
+            raise ValueError("No peers to get info from")
+        coin_state_list: List[CoinState] = await self.service.wallet_state_manager.wallet_node.get_coin_state(
+            [coin_id], peer=peer
+        )
+        if coin_state_list is None or len(coin_state_list) < 1:
+            return {"success": False, "error": f"Coin record 0x{coin_id.hex()} not found"}
+        coin_state: CoinState = coin_state_list[0]
```
Comment (paninaro) [inline chia/rpc/wallet_rpc_api.py:1707]: This chunk of code looks identical to the same functionality found in `nft_get_info`. Can this be refactored into a common method that both RPCs can use?

Maybe there's more of this RPC and `nft_get_info` that can be refactored?

File: chia/rpc/wallet_rpc_api.py [hunk 2]
```diff
@@ -1685,6 +1687,74 @@ async def did_message_spend(self, request) -> EndpointResult:
...
+                        odd_coin += 1
+                    if odd_coin > 1:
+                        return {"success": False, "error": "This is not a singleton, multiple children coins found."}
+                if odd_coin == 0:
+                    return {"success": False, "error": "Cannot find child coin, please wait then retry."}
+                coin_state = coin_state_list[0]
+        # Get parent coin
+        parent_coin_state_list: List[CoinState] = await self.service.wallet_state_manager.wallet_node.get_coin_state(
+            [coin_state.coin.parent_coin_info], peer=peer
+        )
```
Comment (paninaro) [inline chia/rpc/wallet_rpc_api.py:1726]: This chunk of code looks identical to the same functionality found in `nft_get_info`. Can this be refactored into a common method that both RPCs can use?

Maybe there's more of this RPC and `nft_get_info` that can be refactored?

File: chia/rpc/wallet_rpc_api.py [hunk 3]
```diff
@@ -1685,6 +1687,74 @@ async def did_message_spend(self, request) -> EndpointResult:
...
+        if curried_args is None:
+            return {"success": False, "error": "The coin is not a DID."}
+        p2_puzzle, recovery_list_hash, num_verification, singleton_struct, metadata = curried_args
+        uncurried_p2 = uncurry_puzzle(p2_puzzle)
+        (public_key,) = uncurried_p2.args.as_iter()
+
+        return {
+            "success": True,
+            "latest_coin": parent_coin_state.coin,
+            "p2_address": encode_puzzle_hash(p2_puzzle.get_tree_hash(), "xch"),
```
Comment (paninaro) [inline chia/rpc/wallet_rpc_api.py:1749]:
```suggestion
            "p2_address": encode_puzzle_hash(p2_puzzle.get_tree_hash(), AddressType.XCH.hrp(self.service.config)),
```

File: tests/wallet/did_wallet/test_did.py [hunk 4]
```diff
@@ -768,6 +768,65 @@ async def test_update_recovery_list(self, two_wallet_nodes, trusted):
...
+        )
+        spend_bundle = spend_bundle_list[0].spend_bundle
+        await time_out_assert_not_none(5, full_node_api.full_node.mempool_manager.get_spendbundle, spend_bundle.name())
+        for i in range(1, num_blocks):
+            await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(ph1))
+        await time_out_assert(15, did_wallet_1.get_confirmed_balance, 101)
+        await time_out_assert(15, did_wallet_1.get_unconfirmed_balance, 101)
+        response = await api_0.did_get_info({"coin_id": did_wallet_1.did_info.origin_coin.name().hex()})
+        assert response["launcher_id"] == did_wallet_1.did_info.origin_coin.name().hex()
+
```
Comment (paninaro) [inline tests/wallet/did_wallet/test_did.py:829]: Check other fields in the response as well?

**Review by paninaro — APPROVED** [review]
lgtm

**Review by trepca — APPROVED** [review]

**Pull request merged**